lumenlearning/lti-middleware
# DL3-41 sync lineitems

## Description
- Added the ability to sync line items upon any LTIResourceLink request (standard launch) for a new LTI context or an LTI context which is marked as being out of sync.
- Added code for demo mode to insert 150 links into an LMS course with lineitemsSynced set to false. This allowed for local testing while the Valkyrie endpoint(s) were still being developed.
- Renamed the class `LTIToken` to `LTIAdvantageToken` to prevent it from being confused with the `id_token` since the `LTIAdvantageToken` actually represents a bearer token strictly for permitting usage of the LTI Advantage services, not for providing user information upon their login to the tool from the LMS.

### Motivation and Context
Harmony/Valkyrie needs to be aware of the lineitems that exist for a course before users complete any assignments so that Valkyrie knows what lineitem URL corresponds to which assignment when it is completed in order to be able to post the score later on. This will enable students to take quizzes from inside Study Plans instead of having to exit the study plan and enter a separate link for the quiz.

### Fixes
[DL3-41](https://lumenlearning.atlassian.net/browse/DL3-41)

## How Has This Been Tested?
0. As a prerequisite, ensure that the tool has been registered in the LMS and that there exists a course in the LMS that has access to the tool with deep linking enabled. Also ensure that the logging level is set to DEBUG on the lti-middleware.
1. Create a new module and click on the Lumen tool to begin the Deep Linking flow.
2. Select a course from the deep linking menu that has sub-topics to add to the LMS (e.g. Intro to Psychology).
3. Click the Add Course button.
4. Wait 1-2 minutes for the links to be added to the course.
5. Click on one of the new links.
6. Ensure that the LTI 1.3 launch completes properly. In the logs if DEBUG mode is on, it should be possible to see that the calls were made to fetch the lineitems from the LMS and send them to Valkyrie.
7. Return to the LMS module menu and click on another link.
8. Ensure that the LTI 1.3 launch completes properly. In the logs if DEBUG mode is on, it should be possible to see that the criteria to sync lineitems was not met (because they'd already been synced) so the calls to fetch the lineitems from the LMS and send them to Valkyrie were not made.
9. In the LMS, create a copy of the course that you used to do the above testing.
10. Ensure that the new course you just created has access to the Lumen LTI tool in the LMS (org units set up).
11. Navigate to the main content page for this copied course.
12. Click on one of the links in this course.
Ensure that the LTI 1.3 launch completes properly. In the logs if DEBUG mode is on, it should be possible to see that the calls were made to fetch the lineitems from the LMS and send them to Valkyrie.
13. Return to the LMS module menu and click on another link.
14. Ensure that the LTI 1.3 launch completes properly. In the logs if DEBUG mode is on, it should be possible to see that the criteria to sync lineitems was not met (because they'd already been synced) so the calls to fetch the lineitems from the LMS and send them to Valkyrie were not made.

## Screenshots
N/A

---
## Database changes
N/A

## ⚠️ Deployment instructions ⚠️
N/A

## New ENV variables
N/A

---

## Checklist
